aggregator_api/src/lib.rs

```diff
@@ -293,7 +295,32 @@ async fn post_task<C: Clock>(
 
     ds.run_tx_with_name("post_task", |tx| {
         let task = Arc::clone(&task);
-        Box::pin(async move { tx.put_task(&task).await })
+        Box::pin(async move {
+            if let Some(existing_task) = tx.get_task(task.id()).await? {
+            // Check whether the existing task in the DB corresponds to the incoming task, ignoring
+            // those fields that are randomly generated.
+            if existing_task.aggregator_endpoints() == task.aggregator_endpoints()
+                && existing_task.query_type() == task.query_type()
+                && existing_task.vdaf() == task.vdaf()
+                && existing_task.vdaf_verify_keys() == task.vdaf_verify_keys()
+                && existing_task.role() == task.role()
+                && existing_task.max_batch_query_count() == task.max_batch_query_count()
+                && existing_task.task_expiration() == task.task_expiration()
+                && existing_task.min_batch_size() == task.min_batch_size()
+                && existing_task.time_precision() == task.time_precision()
+                && existing_task.collector_hpke_config() == task.collector_hpke_config() {
+                    return Ok(())
+                }
+
+                let err = Error::new(
+                    "task with same VDAF verify key and task ID already exists with different parameters".to_string(),
+                    Status::Conflict,
+                );
+                return Err(datastore::Error::User(err.into()));
```

Comment:
We need to match on the error received by map_err below for `datastore::Error::User`, try downcasting to this module's error type, and return that unwrapped error instead if it succeeds. Otherwise, duplicate tasks would just get the "Error storing task" message.

Reply:
Good catch, I forgot to write a test for the task mutation case.

Reply:
Fixed: I did this to avoid writing identical `match` arms for the `_` case and the case where `datastore::Error:User` wraps something besides `aggregator_api::Error`. Too clever?